upgrade-tests/src/test/java/io/zeebe/test/RollingUpdateTest.java

```diff
@@ -47,47 +52,128 @@ public void setup() {
+      waitUntilBrokerIsRemovedFromTopology(client);
+      zeebeBrokerContainer.start();
+      waitUntilTopologyIsComplete(client);
+    }
   }
 
   @Test
-  public void shouldBeAbleToRestartContainerWithNewVersion() {
+  public void shouldPerformRollingUpgrade() throws InterruptedException {
     // given
-    final var index = 0;
-    final var newVersion = CURRENT_VERSION;
+    final var process =
+        Bpmn.createExecutableProcess("process")
+            .startEvent()
+            .serviceTask("task", s -> s.zeebeJobType("type"))
+            .endEvent()
+            .done();
     Startables.deepStart(containers).join();
-    containers.get(index).shutdownGracefully(Duration.ofSeconds(30));
 
     // when
-    final var zeebeBrokerContainer = replaceBrokerContainer(index, newVersion);
+    createProcessInstance(process);
+    performRollingUpgrade();
 
     // then
-    zeebeBrokerContainer.start();
+    final var latch = new CountDownLatch(1);
+    try (final var client = newZeebeClient(containers.get(0))) {
+      client.newWorker().jobType("type").handler((jobClient, job) -> latch.countDown()).open();
+      assertThat(latch.await(5, TimeUnit.SECONDS)).isTrue();
+    }
+  }
+
+  /**
+   * This method will simulate a rolling upgrade in the same Kubernetes would. It traverses the list
+   * of brokers in reverse order, and shuts them down one by one. After shutting down one node, it
+   * will restart it with the new version, then query another broker's embedded gateway until the
+   * topology is complete. This ensures the restarted broker has not only started but has properly
+   * joined the cluster.
+   */
+  private void performRollingUpgrade() {
+    var availableBroker = containers.get(0);
+    for (int i = containers.size() - 1; i >= 0; i--) {
+      try (final var client = newZeebeClient(availableBroker)) {
+        var container = containers.get(i);
+        container.shutdownGracefully(Duration.ofSeconds(30));
+        waitUntilBrokerIsRemovedFromTopology(client);
+
+        container = upgradeBroker(i);
+        container.start();
+        waitUntilTopologyIsComplete(client);
+
+        availableBroker = container;
+      }
+    }
+  }
+
+  private void createProcessInstance(final BpmnModelInstance process) {
+    try (final var client = newZeebeClient(containers.get(0))) {
+      client
+          .newDeployCommand()
+          .addWorkflowModel(process, "process.bpmn")
+          .send()
+          .join(5, TimeUnit.SECONDS);
+      client
+          .newCreateInstanceCommand()
+          .bpmnProcessId("process")
+          .latestVersion()
+          .variables(Map.of("foo", "bar"))
+          .send()
+          .join();
+    }
+  }
+
+  private void assertTopologyIsComplete(final ZeebeClient zeebeClient, final int clusterSize) {
+    final var topology = zeebeClient.newTopologyRequest().send().join();
+    TopologyAssert.assertThat(topology).isComplete(clusterSize, 1);
```

Comment:
We probably also want to assert the version right?

Reply:
That's a good idea, yeah :+1: